### PR TITLE
*: Support the statement of `rename table`

### LINKED ISSUE #2265: Support the statement of `rename table`
This method can to move tables from one database to a different one.

See [URL](http://dev.mysql.com/doc/refman/5.7/en/rename-table.html)

### PR DESCRIPTION
**Pull request by zimulala**
Fix #2265

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -991,6 +991,38 @@ func (d *ddl) TruncateTable(ctx context.Context, ti ast.Ident) error {
 	return errors.Trace(err)
 }
 
+func (d *ddl) RenameTable(ctx context.Context, oldIdent, newIdent ast.Ident) error {
+	is := d.GetInformationSchema()
+	oldSchema, ok := is.SchemaByName(oldIdent.Schema)
+	if !ok {
+		return errors.Trace(errFileNotFound.GenByArgs(oldIdent.Schema, oldIdent.Name))
```
Comment (shenli) [inline ddl/ddl_api.go]: Why return file not found error?

File: parser/parser.y [hunk 2]
```diff
@@ -4499,6 +4513,7 @@ Statement:
 |	LoadDataStmt
 |	PreparedStmt
 |	RollbackStmt
+|       RenameTableStmt
```
Comment (shenli) [inline parser/parser.y]: Seems not aligned.

File: ddl/table.go [hunk 3]
```diff
@@ -236,3 +214,65 @@ func (d *ddl) onTruncateTable(t *meta.Meta, job *model.Job) error {
...
+	var tableName model.CIStr
+	if err := job.DecodeArgs(&oldSchemaID, &tableName); err != nil {
+		// Invalid arguments, cancel this job.
+		job.State = model.JobCancelled
+		return errors.Trace(err)
+	}
+
+	tblInfo, err := getTableInfo(t, job, oldSchemaID)
+	if err != nil {
+		return errors.Trace(err)
```
Comment (shenli) [inline ddl/table.go]: job.State = model.JobCancelled

File: ddl/table.go [hunk 4]
```diff
@@ -236,3 +214,65 @@ func (d *ddl) onTruncateTable(t *meta.Meta, job *model.Job) error {
...
+	tblInfo.Name = tableName
+	err = t.CreateTable(newSchemaID, tblInfo)
+	if err != nil {
+		job.State = model.JobCancelled
+		return errors.Trace(err)
+	}
+
+	ver, err := updateSchemaVersion(t, job)
+	if err != nil {
+		return errors.Trace(err)
```
Comment (shenli) [inline ddl/table.go]: job.State = model.JobCancelled?

File: ddl/table.go [hunk 5]
```diff
@@ -236,3 +214,65 @@ func (d *ddl) onTruncateTable(t *meta.Meta, job *model.Job) error {
...
+	}
+	job.State = model.JobDone
+	job.BinlogInfo.AddTableInfo(ver, tblInfo)
+	return nil
+}
+
+func checkTableNotExists(t *meta.Meta, job *model.Job, schemaID int64, tableName string) error {
+	// Check this table's database.
+	tables, err := t.ListTables(schemaID)
+	if terror.ErrorEqual(err, meta.ErrDBNotExists) {
```
Comment (shenli) [inline ddl/table.go]: if err != nil {
 if terror.ErrorEqual(err, meta.ErrDBNotExists) {
               job.State = model.JobCancelled
		return errors.Trace(infoschema.ErrDatabaseNotExists)
}
return errors.Trace(err) 
}

File: ddl/ddl_api.go (hunk 1 above)
Comment (zimulala) [inline ddl/ddl_api.go]: To be compatible with MySQL.

File: ddl/table.go (hunk 3 above)
Comment (zimulala) [inline ddl/table.go]: It was done in the function of `getTableInfo`.

File: parser/parser.y (hunk 2 above)
Comment (zimulala) [inline parser/parser.y]: It's OK, you can confirm by clicking on `view`.

File: ddl/table.go (hunk 5 above)
Comment (zimulala) [inline ddl/table.go]: Many other places are like this. If necessary, I will do it in the next PR.

File: ddl/table.go (hunk 4 above)
Comment (zimulala) [inline ddl/table.go]: If it can pass tests, we can ensure that this function doesn't return the error.

File: infoschema/builder.go [hunk 6]
```diff
@@ -65,7 +66,15 @@ func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
 		if oldTableID == newTableID {
 			alloc, _ = b.is.AllocByID(oldTableID)
 		}
-		b.applyDropTable(roDBInfo, oldTableID)
+		if diff.Type == model.ActionRenameTable {
+			oldRoDBInfo, ok := b.is.SchemaByID(diff.OldSchemaID)
+			if !ok {
+				return ErrDatabaseNotExists
+			}
+			b.applyDropTable(oldRoDBInfo, oldTableID)
```
Comment (coocood) [inline infoschema/builder.go]: Need add new table, too?

Comment (zimulala) [inline infoschema/builder.go]: It will add in line 81.

File: infoschema/builder.go [hunk 7]
```diff
@@ -173,21 +153,35 @@ func (b *Builder) applyCreateTable(m *meta.Meta, roDBInfo *model.DBInfo, tableID
...
-	if tableNames, ok := b.is.schemaMap[di.Name.L]; ok {
+	if tableNames, ok := b.is.schemaMap[roDBInfo.Name.L]; ok {
 		delete(tableNames.tables, sortedTables[idx].Meta().Name.L)
 	}
 	// Remove the table in sorted table slice.
 	b.is.sortedTablesBuckets[bucketIdx] = append(sortedTables[0:idx], sortedTables[idx+1:]...)
+
+	// The old DBInfo still holds a reference to old table info, we need to remove it.
+	tables := make([]*model.TableInfo, 0, len(roDBInfo.Tables)-1)
+	for _, tblInfo := range roDBInfo.Tables {
```
Comment (shenli) [inline infoschema/builder.go]: How about find the old table's position first and then roDBInfo.Tables =  append(roDBInfo.Tables[:pos], roDBInfo.Tables[pos+1:])

**Comment by zimulala** [comment]
PTAL @coocood @shenli

File: ddl/ddl_api.go [hunk 8]
```diff
@@ -991,6 +991,38 @@ func (d *ddl) TruncateTable(ctx context.Context, ti ast.Ident) error {
...
+	newSchema, ok := is.SchemaByName(newIdent.Schema)
+	if !ok {
+		return errors.Trace(errErrorOnRename.GenByArgs(oldIdent.Schema, oldIdent.Name,
+			newIdent.Schema, newIdent.Name))
+	}
+	if is.TableExists(newIdent.Schema, newIdent.Name) {
+		return errors.Trace(infoschema.ErrTableExists.GenByArgs(newIdent))
+	}
+
+	job := &model.Job{
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: Would this change affect binlog? Please confirm it. @GregoryIan

**Comment by shenli** [comment]
LGTM

File: ddl/ddl_db_test.go [hunk 9]
```diff
@@ -831,3 +831,37 @@ func (s *testDBSuite) TestTruncateTable(c *C) {
...
+	s.tk.MustExec("create table t (c1 int, c2 int)")
+	s.tk.MustExec("insert t values (1, 1), (2, 2)")
+	ctx := s.tk.Se.(context.Context)
+	is := sessionctx.GetDomain(ctx).InfoSchema()
+	oldTblInfo, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	c.Assert(err, IsNil)
+	oldTblID := oldTblInfo.Meta().ID
+	s.tk.MustExec("create database test1")
+	s.tk.MustExec("use test1")
+	s.tk.MustExec("rename table test.t to test1.t1")
```
Comment (tiancaiamao) [inline ddl/ddl_db_test.go]: Is the table ID  global unique or just unique in each database?

Comment (zimulala) [inline ddl/ddl_db_test.go]: It's global unique.

Comment (tiancaiamao) [inline ddl/ddl_db_test.go]: any test case for renaming tables in the same database?

File: ddl/ddl_api.go (hunk 8 above)
Comment (GregoryIan) [inline ddl/ddl_api.go]: yep,  tidb-binlog needs to make some adjustments to be compatible with it

File: ddl/ddl_db_test.go (hunk 9 above)
Comment (zimulala) [inline ddl/ddl_db_test.go]: Done.

**Pull request merged**